include/nektar_interface/projection/algorithm_types.hpp

```diff
@@ -0,0 +1,269 @@
+                eta0, eta1, data.input[cellx][componant][layerx],
+                &local_mem0[idx_local], &local_mem1[idx_local], stride);
+          }
+
+          idx.barrier(sycl::access::fence_space::local_space);
+
+          auto ndof = Shape::template get_ndof<nmode>();
+          long nthd = idx.get_local_range(1);
+          const auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          auto count =
+              std::min(nthd, std::max(long{0}, npart - layerx + idx_local));
+          auto mode0 = &local_mem0[0];
+          auto mode1 = &local_mem1[0];
+
+          while (idx_local < ndof) {
+            auto temp = Shape::template reduce_dof<nmode, T>(
+                idx_local, count, mode0, mode1, stride);
+            sycl::atomic_ref<T, sycl::memory_order::relaxed,
+                                 sycl::memory_scope::device>
+                coeff_atomic_ref(cell_dof[idx_local]);
+            coeff_atomic_ref.fetch_add(temp);
+            idx_local += nthd;
+          }
+        });
+      });
+    } else {
+      return queue.submit([&](sycl::handler &cgh) {
+        sycl::local_accessor<T> local_mem0{
+            Shape::template local_mem_size<nmode, 0>(stride), cgh};
+        sycl::local_accessor<T> local_mem1{
+            Shape::template local_mem_size<nmode, 1>(stride), cgh};
+        sycl::local_accessor<T> local_mem2{
+            Shape::template local_mem_size<nmode, 2>(stride), cgh};
+
+        cgh.parallel_for<>(range, [=](sycl::nd_item<2> idx) {
+          const int cellx = data.cell_ids[idx.get_global_id(0)];
+          long npart = data.par_per_cell[cellx];
+
+          if (npart == 0)
+            return;
+
+          int idx_local = idx.get_local_id(1);
+          const int layerx = idx.get_global_id(1);
+
+          if (layerx < npart) {
+            auto xi0 = data.positions[cellx][0][layerx];
+            auto xi1 = data.positions[cellx][1][layerx];
+            auto xi2 = data.positions[cellx][2][layerx];
+            T eta0, eta1, eta2;
+            Shape::template loc_coord_to_loc_collapsed<T>(xi0, xi1, xi2, eta0,
+                                                          eta1, eta2);
+            Shape::template fill_local_mem<nmode, T, alpha, beta>(
+                eta0, eta1, eta2, data.input[cellx][componant][layerx],
+                &local_mem0[idx_local], &local_mem1[idx_local],
+                &local_mem2[idx_local], stride);
+          }
+
+          idx.barrier(sycl::access::fence_space::local_space);
+
+          auto ndof = Shape::template get_ndof<nmode>();
+          long nthd = idx.get_local_range(1);
+          const auto cell_dof = &data.dofs[data.dof_offsets[cellx]];
+          auto count =
+              std::min(nthd, std::max(long{0}, npart - layerx + idx_local));
+          auto mode0 = &local_mem0[0];
+          auto mode1 = &local_mem1[0];
+          auto mode2 = &local_mem2[0];
+
+          while (idx_local < ndof) {
```

Comment:
Do we need to parallelise the dofs over local workitems more here? I'm thinking if there is a low number of modes, e.g. 2 then with hexes that is 8 dofs/local works items. However there may be a comparatively large workgroup size, e.g. 256 (depending on how much local memory is required)?

Reply:
This is just based on benchmarks I tried every which way I could think of and this turned out quickest. It would be (maybe) an optimisation to try to fit the local size to the number of dofs - but I haven't tried it
Also it won't actually be slower for e.g. 2 vs 4 (in that loop - the rest is still faste) just not as efficient - and as these are the quickest case anyway I wouldn't worry about it. If you can think of another way then I am happy to benchmark it - but anything using atomics I predict would be slower.